File: lib/domain/field_cultivation/mappers/field_cultivation_sync_apply_mapper.rb
# frozen_string_literal: true

module Domain
  module FieldCultivation
    module Mappers
      module FieldCultivationSyncApplyMapper
        module_function

        # @param plan_snapshot [Dtos::FieldCultivationSyncPlanSnapshot]
        # @param target_snapshot [Dtos::FieldCultivationSyncTargetSnapshot]
        # @return [Dtos::FieldCultivationSyncApply]
        def to_apply(plan_snapshot:, target_snapshot:)
          existing_ids = plan_snapshot.existing_field_cultivation_ids.to_set
          desired_rows = target_snapshot.field_cultivation_rows

          field_cultivations_to_update = desired_rows.select do |row|
            row.field_cultivation_id.present? && existing_ids.include?(row.field_cultivation_id)
          end
          field_cultivations_to_create = desired_rows.reject do |row|
            row.field_cultivation_id.present? && existing_ids.include?(row.field_cultivation_id)
          end
          retained_ids = field_cultivations_to_update.map(&:field_cultivation_id)
          field_cultivation_ids_to_delete = existing_ids.to_a - retained_ids
          cultivation_plan_crop_ids_to_delete =
            FieldCultivationSyncUnreferencedPlanCropIds.ids_to_delete(
              plan_snapshot: plan_snapshot,
              referenced_crop_ids: target_snapshot.referenced_crop_ids
            )

          Dtos::FieldCultivationSyncApply.new(
            field_cultivations_to_update: field_cultivations_to_update,
            field_cultivations_to_create: field_cultivations_to_create,
            field_cultivation_ids_to_delete: field_cultivation_ids_to_delete,
            cultivation_plan_crop_ids_to_delete: cultivation_plan_crop_ids_to_delete,
            cultivation_plan_summary: target_snapshot.cultivation_plan_summary
          )
        end
      end
    end
  end
end

File: crates/agrr-domain/src/field_cultivation/mappers/field_cultivation_sync_apply_mapper.rs
use crate::field_cultivation::dtos::{
    FieldCultivationSyncApply, FieldCultivationSyncPlanSnapshot, FieldCultivationSyncTargetSnapshot,
};
use crate::field_cultivation::mappers::field_cultivation_sync_unreferenced_plan_crop_ids::ids_to_delete;

pub fn to_apply(
    plan_snapshot: &FieldCultivationSyncPlanSnapshot,
    target_snapshot: &FieldCultivationSyncTargetSnapshot,
) -> FieldCultivationSyncApply {
    let existing_ids = plan_snapshot.existing_field_cultivation_ids();
    let desired_rows = &target_snapshot.field_cultivation_rows;

    let field_cultivations_to_update: Vec<_> = desired_rows
        .iter()
        .filter(|row| {
            row.field_cultivation_id
                .is_some_and(|id| existing_ids.contains(&id))
        })
        .cloned()
        .collect();

    let field_cultivations_to_create: Vec<_> = desired_rows
        .iter()
        .filter(|row| {
            !row.field_cultivation_id
                .is_some_and(|id| existing_ids.contains(&id))
        })
        .cloned()
        .collect();

    let retained_ids: Vec<i64> = field_cultivations_to_update
        .iter()
        .filter_map(|row| row.field_cultivation_id)
        .collect();
    let field_cultivation_ids_to_delete: Vec<i64> = existing_ids
        .into_iter()
        .filter(|id| !retained_ids.contains(id))
        .collect();

    let cultivation_plan_crop_ids_to_delete =
        ids_to_delete(plan_snapshot, &target_snapshot.referenced_crop_ids);

    FieldCultivationSyncApply {
        field_cultivations_to_update,
        field_cultivations_to_create,
        field_cultivation_ids_to_delete,
        cultivation_plan_crop_ids_to_delete,
        cultivation_plan_summary: target_snapshot.cultivation_plan_summary.clone(),
    }
}

#[cfg(test)]
mod mappers_field_cultivation_sync_apply_mapper_test_inline {
    use super::*;
    include!(concat!(env!("CARGO_MANIFEST_DIR"), "/test/field_cultivation/mappers_field_cultivation_sync_apply_mapper_test.rs"));
}
